// @salesforce/apex-node/lib/src/streaming/streamingClient.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const faye_1 = require("faye");
const i18n_1 = require("../i18n");
const utils_1 = require("../utils");
const TEST_RESULT_CHANNEL = '/systemTopic/TestResult';
const DEFAULT_STREAMING_TIMEOUT_MS = 14400;
class Deferred {
    constructor() {
        this.promise = new Promise(resolve => (this.resolve = resolve));
    }
}
class StreamingClient {
    constructor(connection, progress) {
        this.apiVersion = '36.0';
        this.subscribedTestRunIdDeferred = new Deferred();
        this.conn = connection;
        this.progress = progress;
        const streamUrl = this.getStreamURL(this.conn.instanceUrl);
        this.client = new faye_1.Client(streamUrl, {
            timeout: DEFAULT_STREAMING_TIMEOUT_MS
        });
        this.client.on('transport:up', () => {
            var _a;
            (_a = this.progress) === null || _a === void 0 ? void 0 : _a.report({
                type: 'StreamingClientProgress',
                value: 'streamingTransportUp',
                message: i18n_1.nls.localize('streamingTransportUp')
            });
        });
        this.client.on('transport:down', () => {
            var _a;
            (_a = this.progress) === null || _a === void 0 ? void 0 : _a.report({
                type: 'StreamingClientProgress',
                value: 'streamingTransportDown',
                message: i18n_1.nls.localize('streamingTransportDown')
            });
        });
        this.client.addExtension({
            incoming: (message, callback) => {
                if (message && message.error) {
                    if (message.channel === '/meta/handshake') {
                        this.client.disconnect();
                        throw new Error(i18n_1.nls.localize('streamingHandshakeFail', message.error));
                    }
                    this.client.disconnect();
                    throw new Error(message.error);
                }
                callback(message);
            }
        });
    }
    get subscribedTestRunIdPromise() {
        return this.subscribedTestRunIdDeferred.promise;
    }
    removeTrailingSlashURL(instanceUrl) {
        return instanceUrl ? instanceUrl.replace(/\/+$/, '') : '';
    }
    getStreamURL(instanceUrl) {
        const urlElements = [
            this.removeTrailingSlashURL(instanceUrl),
            'cometd',
            this.apiVersion
        ];
        return urlElements.join('/');
    }
    // NOTE: There's an intermittent auth issue with Streaming API that requires the connection to be refreshed
    // The builtin org.refreshAuth() util only refreshes the connection associated with the instance of the org you provide, not all connections associated with that username's orgs
    init() {
        return __awaiter(this, void 0, void 0, function* () {
            yield utils_1.refreshAuth(this.conn);
            const accessToken = this.conn.getConnectionOptions().accessToken;
            if (accessToken) {
                this.client.setHeader('Authorization', `OAuth ${accessToken}`);
            }
            else {
                throw new Error(i18n_1.nls.localize('noAccessTokenFound'));
            }
        });
    }
    handshake() {
        return new Promise(resolve => {
            this.client.handshake(() => {
                resolve();
            });
        });
    }
    disconnect() {
        this.client.disconnect();
    }
    subscribe(action) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((subscriptionResolve, subscriptionReject) => {
                try {
                    this.client.subscribe(TEST_RESULT_CHANNEL, (message) => __awaiter(this, void 0, void 0, function* () {
                        const result = yield this.handler(message);
                        if (result) {
                            this.client.disconnect();
                            subscriptionResolve({
                                runId: this.subscribedTestRunId,
                                queueItem: result
                            });
                        }
                    }));
                    action()
                        .then(id => {
                        this.subscribedTestRunId = id;
                        this.subscribedTestRunIdDeferred.resolve(id);
                    })
                        .catch(e => {
                        this.client.disconnect();
                        subscriptionReject(e);
                    });
                }
                catch (e) {
                    this.client.disconnect();
                    subscriptionReject(e);
                }
            });
        });
    }
    isValidTestRunID(testRunId, subscribedId) {
        if (testRunId.length !== 15 && testRunId.length !== 18) {
            return false;
        }
        const testRunId15char = testRunId.substring(0, 14);
        if (subscribedId) {
            const subscribedTestRunId15char = subscribedId.substring(0, 14);
            return subscribedTestRunId15char === testRunId15char;
        }
        return true;
    }
    handler(message, runId) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const testRunId = runId || message.sobject.Id;
            if (!this.isValidTestRunID(testRunId, this.subscribedTestRunId)) {
                return null;
            }
            const result = yield this.getCompletedTestRun(testRunId);
            if (result) {
                return result;
            }
            (_a = this.progress) === null || _a === void 0 ? void 0 : _a.report({
                type: 'StreamingClientProgress',
                value: 'streamingProcessingTestRun',
                message: i18n_1.nls.localize('streamingProcessingTestRun', testRunId),
                testRunId
            });
            return null;
        });
    }
    getCompletedTestRun(testRunId) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const queryApexTestQueueItem = `SELECT Id, Status, ApexClassId, TestRunResultId FROM ApexTestQueueItem WHERE ParentJobId = '${testRunId}'`;
            const result = (yield this.conn.tooling.query(queryApexTestQueueItem));
            if (result.records.length === 0) {
                throw new Error(i18n_1.nls.localize('noTestQueueResults', testRunId));
            }
            (_a = this.progress) === null || _a === void 0 ? void 0 : _a.report({
                type: 'TestQueueProgress',
                value: result
            });
            for (let i = 0; i < result.records.length; i++) {
                const item = result.records[i];
                if (item.Status === "Queued" /* Queued */ ||
                    item.Status === "Holding" /* Holding */ ||
                    item.Status === "Preparing" /* Preparing */ ||
                    item.Status === "Processing" /* Processing */) {
                    return null;
                }
            }
            return result;
        });
    }
}
exports.StreamingClient = StreamingClient;
