// @salesforce/source-deploy-retrieve/lib/src/metadata-registry/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const data = require("./data/registry.json");
const registry_1 = require("../utils/registry");
var metadataResolver_1 = require("./metadataResolver");
exports.MetadataResolver = metadataResolver_1.MetadataResolver;
var registryAccess_1 = require("./registryAccess");
exports.RegistryAccess = registryAccess_1.RegistryAccess;
var manifestGenerator_1 = require("./manifestGenerator");
exports.ManifestGenerator = manifestGenerator_1.ManifestGenerator;
var treeContainers_1 = require("./treeContainers");
exports.BaseTreeContainer = treeContainers_1.BaseTreeContainer;
exports.NodeFSTreeContainer = treeContainers_1.NodeFSTreeContainer;
exports.VirtualTreeContainer = treeContainers_1.VirtualTreeContainer;
exports.ZipTreeContainer = treeContainers_1.ZipTreeContainer;
var sourceComponent_1 = require("./sourceComponent");
exports.SourceComponent = sourceComponent_1.SourceComponent;
var types_1 = require("./types");
exports.DecompositionStrategy = types_1.DecompositionStrategy;
exports.TransformerStrategy = types_1.TransformerStrategy;
var forceIgnore_1 = require("./forceIgnore");
exports.ForceIgnore = forceIgnore_1.ForceIgnore;
/**
 * Direct access to the JSON registry data. Useful for autocompletions.
 */
exports.registryData = registry_1.deepFreeze(data);
