// @salesforce/source-deploy-retrieve/lib/src/metadata-registry/types.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
var DecompositionStrategy;
(function (DecompositionStrategy) {
    /**
     * Elements of child types are decomposed to the same folder the parent object is in
     */
    DecompositionStrategy["TopLevel"] = "topLevel";
    /**
     * Elements of child types are decomposed into folders of their respective types
     */
    DecompositionStrategy["FolderPerType"] = "folderPerType";
})(DecompositionStrategy = exports.DecompositionStrategy || (exports.DecompositionStrategy = {}));
var TransformerStrategy;
(function (TransformerStrategy) {
    TransformerStrategy["Standard"] = "standard";
    TransformerStrategy["Decomposed"] = "decomposed";
    TransformerStrategy["StaticResource"] = "staticResource";
})(TransformerStrategy = exports.TransformerStrategy || (exports.TransformerStrategy = {}));
